ARCs/arc-0018.md

```diff
@@ -0,0 +1,348 @@
+
+Transfers the Asset after checking that the royalty policy is adhered to. This call must be sent by the current owner (the `from` parameter)
+
+The `royalty_asset` is the ASA ID to be transferred. It MUST have a policy defined prior to attempting a transfer.
+
+The `from` parameter is the address for the account the ASA is transferred from.
+
+The `to` parameter is the address for the account the ASA is transferred to.
+
+The `royalty_receiver` parameter is the address for the account that collects the royalty and takes care of any redistribution logic (if any). 
+
+The `royalty_asset_amount` parameter is the number of units of this ASA ID to transfer. The amount MUST be less than or equal to the amount `offered` by the `from` account.
+
+The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address. 
+
+The `payment_asset` parameter is used in the case that the payment is being made with some asset rather than with Algos. It MUST match the Asset ID of the AssetTransfer payment transaction. 
+
+The transfer call SHOULD be part of a group with a size of 2 (app call + payment/asset transfer)
+
+> While its possible to enforce the group size limit, it is possible to circumvent the royalty enforcement logic by simply making an Inner Transaction application call with the appropriate parameters and a small payment, then in the same outer group the "real" payment. The counter-party risk remains the same since the inner transaction is atomic with the outers. 
+> Some ways to prevent a small royalty payment and larger payment in a later transaction of the same group might be by using an `allow` list that is checked against the `auth_addr` of the offer call. The `allow` list would be comprised of known and trusted marketplaces that do not attempt to circumvent the royalty policy. The `allow` list may be implicit as well by transferring a specific asset to the `auth_addr` as frozen and on `offer` a the balance must be > 0 to allow the `auth_addr` to be persisted.
+
+Upon each transfer the the Royalty Enforcer SHOULD assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.
+
+> The rational for this check is ensuring that `Rekeying` is not exploited to transfer the `royalty_asset` ownership circumventing the Royalty Policy. Two parties involved in the trade could create a Group Transactions in which: 1. Seller rekeys the `from` to the Buyer `to`, 2. Buyer sends payment to `other_seller_account`. Note that nothing can prevent the users to execute such trade but the Royalty Enforcer tries to dissuade this behavior: if such trade happens, the Receiver will no longer be able to re-sell the Royalty ASA in the regular way.
```

Comment:
This does not prevent the use of smart contract controlling the account holding the asset.
That being said, such an attack requires all the users to agree on using such a smart contract from the beginning, so I guess this is unlikely to be a real threat.